### PR TITLE
[Calyx/Cider + Frontends] Metadata rewrite

### PR DESCRIPTION
**Pull request by EclecticGriffin**
This implements the proposal in #994, the only major difference with the described minimum proposal is that this uses `#{` and `}#` as the delimiting characters, rather than `{` and `}` as the latter are likely to appear in source positions.

This means that the metadata now appears at the end of the file like:
```
...
METADATA #{ 
 // Some stuff
}#
```

On the IR side the stuff between the brackets is treated as an opaque string so it only gets parsed as the metadata source map when given to Cider.

The main benefit is that this simplifies the user calls, so that the following now suffices to pass source position from the frontends
```
fud e --to debugger tests/frontend/systolic/array-3.systolic -s systolic.flags "--write-metadata " -q
```
```
fud e --to debugger lenet.relay -s relay.flags "--write-metadata " -q
```

If metadata is not present or an empty string it is not printed by the IR printer. 

There are also some minor changes to the language parser to accommodate the optional existence of metadata, mostly because `pest_consume` was fighting me

### CODE REVIEW

File: calyx/src/ir/printer.rs [hunk 1]
```diff
@@ -522,4 +522,13 @@ impl Printer {
             ),
         }
     }
+
+    /// Formats the top-level metadata if present
+    pub fn format_metadata(metadata: &Option<String>) -> String {
```
Comment (rachitnigam) [inline calyx/src/ir/printer.rs:527]: Can you add a call to this in the `main.rs` and add a new test for metadata stuff in `tests/parsing`?

Comment (rachitnigam) [reply calyx/src/ir/printer.rs:527]: Nvm, I just didn't see it!

File: calyx/src/ir/printer.rs [hunk 2]
```diff
@@ -522,4 +522,13 @@ impl Printer {
...
+        if let Some(metadata_str) = metadata {
+            format!("METADATA #{{\n{}\n}}#\n", metadata_str)
```
Comment (rachitnigam) [inline calyx/src/ir/printer.rs:529]: dumb nit, can we make this `metadata` instead? All other keywords are small caps.

File: calyx-py/calyx/py_ast.py [hunk 3]
```diff
@@ -355,6 +355,18 @@ def doc(self) -> str:
         return block(cond, true_branch, sep="") + false_branch
 
 
+@dataclass
+class Metadata:
```
Comment (sampsyo) [inline calyx-py/calyx/py_ast.py:359]: Rad that this is even in the Python AST module… super convenient.

**Review by sampsyo — COMMENTED** [review]
Wahoo; looks just great from here! I don't have a strong opinion on the whole caps thing.

File: calyx/src/ir/printer.rs (hunk 2 above)
Comment (EclecticGriffin) [reply calyx/src/ir/printer.rs:529]: technically the metadata label is case insensitive, so we can change the default printing pretty easily

Comment (EclecticGriffin) [reply calyx/src/ir/printer.rs]: lower-cased!

**Comment by rachitnigam** [comment]
LGTM!

**Comment by EclecticGriffin** [comment]
@cgyurgyik just a note that the metadata format changed slightly from the original proposal

**Pull request merged**